### PR TITLE
Cancellable promises

### PR DESCRIPTION
**Pull request by jkirkpatrick24**
# Summary

- Cancels Promises if the component has unmounted.
- Refactors to use hooks internally as opposed to component state.
- Upgrade dependencies.

# Test Plan

Ensure all the unit tests are passing.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,24 +19,32 @@
   "homepage": "https://github.com/eadmundo/react-resolve-dependencies",
   "license": "MIT",
   "dependencies": {
-    "react": "^15.5.4",
-    "react-dom": "^15.5.4"
+    "@babel/runtime": "^7.5.5",
```
Comment (jkirkpatrick24) [inline package.json]: Oops. I think this can be removed.

File: package.json [hunk 2]
```diff
@@ -26,24 +19,32 @@
...
-    "jsdom": "^9.8.3",
-    "mocha": "^2.4.5",
-    "react-addons-test-utils": "^15.5.1",
-    "rimraf": "^2.5.2",
-    "sinon": "^1.17.6",
-    "watch": "^0.17.1"
+    "@babel/cli": "^7.5.5",
+    "@babel/core": "^7.5.5",
+    "@babel/plugin-proposal-class-properties": "^7.5.5",
+    "@babel/plugin-transform-runtime": "^7.5.5",
```
Comment (jkirkpatrick24) [inline package.json]: Same here. Leftover

File: src/index.js [hunk 3]
```diff
@@ -1,99 +1,83 @@
...
-          WrappedComponent.onDependencyFailure(reason)
+      };
+
+      const onDependencyFailure = reason => {
+        const { isCanceled } = reason;
+
+        if (!isCanceled) {
+          // If we have cancelled the promise when unmounting the component, don't
+          // attempt to update any state values.
+          setDependenciesFailed(true);
```
Comment (jkirkpatrick24) [inline src/index.js:46]: These state updates were causing the memory leaks before. Now if we unmount, always reject the promise so we call onDependencyFailure and if the reason is cancellation (which happens when we unmount), skip all additional callbacks and state update actions.

File: test/resolveDependenciesTests.js [hunk 4]
```diff
@@ -87,21 +62,11 @@ describe('resolveDependencies', () => {
     })
 
     it('should render the FailureComponent', () => {
-      expect(mountedComponent.contains(<FailureComponent />)).to.be.true;
-    })
-
-    it('should set dependenciesFailed to true', () => {
-      expect(mountedComponent.state('dependenciesFailed')).to.be.true;
+      expect(mountedComponent.update().find('.failed').length === 1);
```
Comment (jkirkpatrick24) [inline test/resolveDependenciesTests.js:65]: Needed to call `.update()` here to re-render the component.

File: test/resolveDependenciesTests.js [hunk 5]
```diff
@@ -135,14 +96,39 @@ describe('resolveDependencies', () => {
...
       expect(mountedComponent.contains(<WrappedComponent />)).to.be.true;
     })
+  })
 
-    it('should set dependenciesResolved to true', () => {
-      expect(mountedComponent.state('dependenciesResolved')).to.be.true;
-    })
+  context('cancelling dependency resolution', () => {
+    it('should not call onDependencyResolution or onDependencyFailure', (done) => {
+      // Promise will never resolve
```
Comment (jkirkpatrick24) [inline test/resolveDependenciesTests.js:103]:
```suggestion
```

**Review by eadmundo — APPROVED** [review]
Looking good!

**Pull request merged**